examples/imagenet/ilsvrc12.h

```diff
@@ -0,0 +1,409 @@
+
+  return image;
+}
+
+void ILSVRC::Write(string outfile, singa::io::Mode mode) {
+  BinFileWriter bfwriter;
+  if (buff.size() > 0) {
+    bfwriter.Open(outfile, mode);
+    for (auto sample : buff) {
+      bfwriter.Write(outfile, sample);
+    }
+    bfwriter.Flush();
+    bfwriter.Close();
+    std::vector<string>().swap(buff);
+  }
+}
+
+void ILSVRC::WriteMean(Tensor& mean, string path) {
+  Tensor mean_lb(Shape{1}, kInt);
+  std::vector<Tensor> input;
+  input.push_back(mean);
+  input.push_back(mean_lb);
+  BinFileWriter bfwriter;
+  bfwriter.Open(path, kCreate);
+  bfwriter.Write(path, encoder->Encode(input));
+  bfwriter.Flush();
+  bfwriter.Close();
+}
+
+void ILSVRC::CreateTrainData(string image_list, string input_folder,
+                             string output_folder, size_t file_size = 12800) {
+  std::vector<std::pair<string, int>> file_list;
+  size_t *sum = new size_t[kImageNBytes];
+  for (size_t i = 0; i < kImageNBytes; i++)
+    sum[i] = 0u;
+  string image_file_name;
+  int label;
+  std::ifstream image_list_file(image_list.c_str(), std::ios::in);
+  while (image_list_file >> image_file_name >> label)
+    file_list.push_back(std::make_pair(image_file_name, label));
+  LOG(INFO) << "Data Shuffling";
+  std::shuffle(file_list.begin(), file_list.end(),
+               std::default_random_engine());
+  LOG(INFO) << "Total number of training images is " << file_list.size();
+  num_train_images = file_list.size();
+  num_train_images = 12900;
+  if (file_size == 0) file_size = num_train_images;
+  for (size_t imageid = 0; imageid < num_train_images; imageid++) {
+    string path = input_folder + "/" + file_list[imageid].first;
+    Tensor image = ReadImage(path);
+    auto image_data = image.data<unsigned char>();
+    for (size_t i = 0; i < kImageNBytes; i++)
+      sum[i] += image_data[i];
+    label = file_list[imageid].second;
+    Tensor lb(Shape{1}, kInt);
+    lb.CopyDataFromHostPtr<int>(&label, 1);
+    std::vector<Tensor> input;
+    input.push_back(image);
+    input.push_back(lb);
+    string encode_str = encoder->Encode(input);
+    buff.push_back(encode_str);
+    if ((imageid + 1) % file_size == 0) {
+      string file = output_folder + "/train" +
+                    std::to_string((imageid + 1) / file_size) + ".bin";
+      int count = buff.size();
+      Write(file, kCreate);
+      LOG(INFO) << "Write " << count << " images into " << file;
+    }
+  }
+  // if buff is not empty, write to another file.
+  if (buff.size() > 0) {
+    string file = output_folder + "/train" +
+                  std::to_string(num_train_images / file_size + 1) + ".bin";
+    int count = buff.size();
+    Write(file, kCreate);
+    LOG(INFO) << "Write " << count << " images into " << file;
+  }
+  Tensor mean = Tensor(Shape{3, kImageSize, kImageSize}, kUChar);
+  unsigned char *mean_data = new unsigned char[kImageNBytes];
+  for (size_t i = 0 ; i < kImageNBytes; i++)
+    mean_data[i] = (unsigned char)(sum[i] / num_train_images);
```

Comment:
typically type of mean is float.
it may not be a big problem if use unsigned char to store mean..